aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -96,206 +127,442 @@ impl AggregatorID {
 ///
 /// TODO: while we support tracking of the history, it is not yet fully used on
 /// executor side because we don't know how to throw errors.
-#[derive(Debug)]
-pub struct History {
-    pub max_positive: u128,
-    pub min_negative: u128,
+#[derive(Clone, Hash, Copy, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub struct DeltaHistory {
+    pub max_achieved_positive_delta: u128,
+    pub min_achieved_negative_delta: u128,
+    // `min_overflow_positive_delta` is None in two possible cases:
+    // 1. No overflow occured in the try_add/try_sub functions throughout the
+    // transaction execution.
+    // 2. The only overflows that occured in the try_add/try_sub functions in
+    // this transaction execution are with delta that exceeds limit.
+    pub min_overflow_positive_delta: Option<u128>,
+    // `max_underflow_negative_delta` is None in two possible cases:
+    // 1. No underflow occured in the try_add/try_sub functions throughout the
+    // transaction execution.
+    // 2. The only underflows that occured in the try_add/try_sub functions in
+    // this transaction execution are with delta that drops below -limit.
+    pub max_underflow_negative_delta: Option<u128>,
+}
+
+impl Default for DeltaHistory {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
-impl History {
-    fn new() -> Self {
-        History {
-            max_positive: 0,
-            min_negative: 0,
+impl DeltaHistory {
+    pub fn new() -> Self {
+        DeltaHistory {
+            max_achieved_positive_delta: 0,
+            min_achieved_negative_delta: 0,
+            min_overflow_positive_delta: None,
+            max_underflow_negative_delta: None,
         }
     }
 
-    fn record_positive(&mut self, value: u128) {
-        self.max_positive = u128::max(self.max_positive, value);
+    /// Records observed delta in history. Should be called after an operation (addition/subtraction)
+    /// is successful to record its side-effects.
+    fn record_success(&mut self, delta: DeltaValue) {
+        match delta {
+            DeltaValue::Positive(value) => {
+                self.max_achieved_positive_delta =
+                    u128::max(self.max_achieved_positive_delta, value)
+            },
+            DeltaValue::Negative(value) => {
+                self.min_achieved_negative_delta =
+                    u128::max(self.min_achieved_negative_delta, value)
+            },
+        }
     }
 
-    fn record_negative(&mut self, value: u128) {
-        self.min_negative = u128::max(self.min_negative, value);
+    /// Records overflows in history. Should be called after an addition is unsuccessful
```

Comment:
This and the next one an be a single method, i.e. record_failure to mirror record_success, on &mut u128, and the caller can apply it to min_overflow_positive_delta and max_underflow_negative_delta members.